src/util/timer.hpp

```diff
@@ -1,170 +1,58 @@
+
+      if (!this->children.empty()) {
+        auto children = nlohmann::json::array();
+        for (auto const& timer_data : this->children) {
+          children.push_back(timer_data.to_json());
+        }
+        obj["children"] = children;
+      }
+
+      return obj;
+    }  
   };
 
-  /// A pointer to a timer, which stops and pops the timer upon destruction.
-  ///
-  /// This is useful for scope timers, and must always be moved from
   struct ScopedTimer {
-
-    /// Move only
-    ScopedTimer(const ScopedTimer& rhs) = delete;
-
-    /// Move construct this handler
-    ///
-    /// \postconditions `this->timer == rhs.timer` and also,
-    /// `rhs.timer == nullptr`
-    ScopedTimer(ScopedTimer&& rhs) noexcept;
-
-    /// Destruct the handler and stop the timer.
-    ///
-    /// \effects if `timer != nullptr`, call `stop(*this)`.
-    /// Otherwise, do nothing
+    ScopedTimer(timer_id id);
     ~ScopedTimer();
 
-  private:
-
-    /// Construct a handle to the provided timer
-    ///
-    /// \postconditions `this->timer == &timer`
-    explicit ScopedTimer(Timer& timer) noexcept;
-
-    /// Construct a handle to the provided timer
-    ///
-    /// \note `timer` can be null
-    /// \postconditions `this->timer == timer`
-    explicit ScopedTimer(Timer* timer) noexcept;
-
-    Timer* timer;
-
-    friend ScopedTimer start_scoped(timer_id);
-    friend ScopedTimer start_scoped(Timer&);
+    private:
+      timer_id id;
+      time_point start;
+      std::vector<TimerData> children;
+      ScopedTimer *parent;
   };
 
-  /// Find or create a child of the current top of the stack, which matches id
-  /// `id`.
-  Timer& find_or_make(timer_id id);
-
   /// Start a timer with id `id`, and push it to the stack.
-  Timer& start(timer_id id);
-
-  /// Start `timer` and push it onto the stack.
-  Timer& start(Timer& timer);
-
-  /// Start a timer with id `id`, and push it to the stack.
-  ScopedTimer start_scoped(timer_id id);
-
-  /// Start the timer pointed to by `TimerHandle`
-  ScopedTimer start_scoped(Timer&);
-
-  /// Stop the current timer, and pop it off the stack
-  void stop();
-
-  /// Stop the timer, and pop it off the stack if it is there.
-  ///
-  /// Currently unimplemented
-  /// \effects `timer.stop()` and, if `&timer` is in `timer_stack`, stop and pop
-  /// it and everything after it.
-  void stop(Timer& timer) = delete;
+  ScopedTimer start(timer_id id);
 
-  /// Tick a timer with id `id`, and push it to the back of the stack if it is
-  /// not there already.
-  Timer& tick(timer_id id);
-
-  /// Tick `timer`
-  ///
-  /// \effects `timer.tick()` and, if `timer_stack.back() != &timer`, push it to
-  /// the stack.
-  Timer& tick(Timer& timer);
-
-  nlohmann::json serialize();
-
-} // otto::util::timer
-
-// M A C R O S /////////////////////////////////////////////////////////////////
-
-#if OTTO_ENABLE_TIMERS
-#define TIME_SCOPE(name) auto scope_timer = ::otto::util::timer::start_scoped(name);
```

Comment:
This wont work. To time a scope, you need to store the ScopeTimer in a variable, so it is RAII-destructed when it leaves scope. Thats what this macro is for

Reply:
This actually works fine.  The returned ScopedTimer isn't destroyed until after its enclosing scope is closed.

Reply:
After our conversation, and some more informed and detailed testing, you're absolutely right.

I've added [[nodiscard]], which now generates warnings about unused instances, and I've brought back the TIME_SCOPE macro for most use cases.